### PR TITLE
GenericPowerMeter: Add support for calling specialized power meter implementations

### PR DESCRIPTION
**Pull request by dorezyuk**
This pr adds support for transactions in the generic power meter. 

The mechanics of transactions differ significantly between power meter vendors. Therefore we can't just extend the config in the same way we did for reading power meter values. This pr adds an optional configuration field which will delegate the transaction calls to a "specific" power meter. This allows the user to re-use the work done by the generic power meter and still have transactions.

### CODE REVIEW

File: modules/GenericPowermeter/main/powermeterImpl.cpp [hunk 1]
```diff
@@ -48,15 +48,19 @@ void powermeterImpl::ready() {
...
+    return mod->r_specific_power_meter.front()->call_stop_transaction(transaction_id);
 };
 
 types::powermeter::TransactionStartResponse
 powermeterImpl::handle_start_transaction(types::powermeter::TransactionReq& value) {
-    return {types::powermeter::TransactionRequestStatus::NOT_SUPPORTED,
-            "Generic powermeter does not support the start_transaction command"};
+    if (mod->r_specific_power_meter.empty())
+        return {types::powermeter::TransactionRequestStatus::NOT_SUPPORTED,
+                "Generic powermeter does not support the start_transaction command"};
```
Comment (SirVer) [inline modules/GenericPowermeter/main/powermeterImpl.cpp:62]: same comment

File: modules/GenericPowermeter/manifest.yaml [hunk 2]
```diff
@@ -23,6 +23,10 @@ provides:
 requires:
   serial_comm_hub:
     interface: serial_communication_hub
+  specific_power_meter:
```
Comment (SirVer) [inline modules/GenericPowermeter/manifest.yaml:26]: This needs a comment. It is not obvious why this is useful to a first time reader. Maybe the `implementation_id` could also be more specific. Something like "start_transaction_meter"? I do not love this either, but I have no better idea at the moment.

File: modules/GenericPowermeter/serialization.hpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Qwello GmbH and Contributors to EVerest
+
+/// @brief Serialization and deserialization for the modbus interface.
+///
+/// The format has a major quirk: Even though our exchanged data is four bytes
```
Comment (SirVer) [inline modules/GenericPowermeter/serialization.hpp:6]: can you add a static_assert for this somehow?

File: modules/GenericPowermeter/serialization.hpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+inline void check_message(const Result& _message) {
+    if (!_message.value.has_value())
+        throw std::invalid_argument("The `value` must be set");
+}
+
+/// @brief Deserialization for small integral types (less or equal to two bytes).
+/// @param _message The message read.
+/// @param _result The result which will contain the deserialized data.
+template <typename Output, std::enable_if_t<std::is_integral<Output>::value && (sizeof(Output) <= 2), bool> = true>
+inline void deserialize(const Result& _message, Output& _result) {
```
Comment (SirVer) [inline modules/GenericPowermeter/serialization.hpp:38]: why are you using `_` at the beginning of your variables? For one, that is confusing to read, for two in c++:

> Each identifier that begins with an underscore is reserved to the implementation for use as a name in the global namespace.

From: https://eel.is/c++draft/lex.name#3.2.

So you should never start your identfiers with an underscore - unless you implement a c++ standard library or a c++ compiler. Same comment for all variables that start with `_` in your code here.

File: modules/GenericPowermeter/serialization.hpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (SirVer) [inline modules/GenericPowermeter/serialization.hpp:1]: I see that this code is tested, but where is it used?

Comment (dorezyuk) [reply modules/GenericPowermeter/serialization.hpp:1]: Its (for now) used in https://github.com/EVerest/everest-core/pull/443/files

File: modules/GenericPowermeter/serialization.hpp (hunk 3 above)
Comment (dorezyuk) [reply modules/GenericPowermeter/serialization.hpp:6]: Would be nice but I don't think so.. its not expressed through the type system

File: modules/GenericPowermeter/serialization.hpp (hunk 5 above)
Comment (SirVer) [reply modules/GenericPowermeter/serialization.hpp:1]: Ok, fine by me.

File: modules/GenericPowermeter/manifest.yaml (hunk 2 above)
Comment (SirVer) [reply modules/GenericPowermeter/manifest.yaml:26]: friendly ping on this.

File: modules/GenericPowermeter/main/powermeterImpl.cpp [hunk 6]
```diff
@@ -48,15 +48,19 @@ void powermeterImpl::ready() {
 }
 
 types::powermeter::TransactionStopResponse powermeterImpl::handle_stop_transaction(std::string& transaction_id) {
-    return {types::powermeter::TransactionRequestStatus::NOT_SUPPORTED,
-            {},
-            "Generic powermeter does not support the stop_transaction command"};
+    if (mod->r_transactional_power_meter.empty())
+        return {types::powermeter::TransactionRequestStatus::NOT_SUPPORTED,
+                {},
+                "No specific powermeter configured to start a transaction"};
```
Comment (barsnick) [inline modules/GenericPowermeter/main/powermeterImpl.cpp:54]: This should be "No specific powermeter configured to **stop** a transaction"

**Comment by corneliusclaussen** [comment]
See feedback on https://github.com/EVerest/everest-core/pull/443

**Comment by corneliusclaussen** [comment]
Can we close this? I think this should be solved differently

**Pull request closed without merging**